### ISSUE TITLE
Worker takes 100% of CPU

### ISSUE THREAD

**Issue by BlackYoup** [sozu-proxy/sozu#266]
From time to time, I have a worker that takes 100% of the CPU. No particular logs displayed but it is stuck in the `<sozu_lib::network::session::Session<ServerConfiguration, Client>>::ready ()` loop

I don't have a reproduction case for now

**Comment by Geal** [comment]
to debug this, you can decomment these lines:
- 494: `//info!("PROXY\t{:?} got events: {:?}", token, events);`
- 565: `//info!("PROXY\t{:?} {:?} | {:?} front: {:?} | back: {:?} ", client_token, events, self.clients[client_token].readiness(), front_interest, back_interest);`

**Comment by BlackYoup** [comment]
Just reproduced the problem after a few minutes of uptime, logs were only:
```
2017-11-30T22:47:00Z 1710484068839      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                 
2017-11-30T22:47:00Z 1710484093769      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484117869      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484141609      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484208048      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484234938      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484260438      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484285988      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484310818      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484335137      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
2017-11-30T22:47:00Z 1710484360327      16617 TAG-6 INFO        PROXY   FrontToken(0) Readable | Writable | Hup | Readiness { front_interest: Writable | Error | Hup, back_interest: Error | Hup, front_readiness: (empty), back_readiness: Readable | Writable | Hup } front: (empty) | back: Hup                                                                                                  
```

**Comment by BlackYoup** [comment]
A bit of information. I managed to reproduce (not on purpose) the problem 3 times with the last time in a matter of seconds.

Here are sozu logs and tcpdump logs: https://blackyoup.cellar.services.clever-cloud.com/sozu-logs-issue-266.tar.gz
There is a `#START` line when the problem starts (at `21:36:56` for sozu and `22:36:56` for tcpdump). I can't understand yet what are the exact lines in tcpdump logs when the problem occurs, maybe you'll be able to tell me. 

What I've found so far is that the backend becomes `HUP` at some point. So we enter in that condition: https://github.com/sozu-proxy/sozu/blob/22787262b2ee2364b6a34615cb28fdb7b96d6ff7/lib/src/network/session.rs#L606

The `back_hup()` right after returns `ClientResult:Continue`. In that method, there is this condition: https://github.com/sozu-proxy/sozu/blob/22787262b2ee2364b6a34615cb28fdb7b96d6ff7/lib/src/network/protocol/http.rs#L264

Obviously, the `output_data_size()` returns a non zero length, maybe because the backend become `HUP` before it could send all of the expected data.
What I have more troubles to understand is `next_output_data()` and what it returns. In my case, I guess it returns a non-zero length which makes the condition fail and not closing the backend.

From there, the loop goes indefinitely
